File: packages/core/src/seed.ts
/**
 * Seed helper — composes the most stable identifier from a user-shaped
 * object into a single string for `createAvatar`.
 *
 * Priority: `id` → `email` → `name + createdAt` → `name` alone.
 *
 * The whole point: stop devs accidentally passing a display name. Names
 * collide; ids and emails don't. When only a name is available, composing
 * it with `createdAt` makes the result globally unique while remaining
 * stable across renders (assuming `createdAt` is set once at signup).
 */

import { sha256Hex } from './sha256.js';

/**
 * Normalize an email the same way Gravatar does — trim + lowercase, NFC.
 * Exported so callers can reproduce the canonical form before hashing.
 */
export function normalizeEmail(email: string): string {
  return email.trim().toLowerCase().normalize('NFC');
}

/**
 * Turn an email into a stable, opaque seed using Gravatar's scheme:
 * `sha256(trim(lowercase(email)))` → lowercase hex.
 *
 * Why: passing raw emails as seeds leaks them through URLs (server logs,
 * Referer headers, browser history, CDN cache keys, analytics). The hash
 * is stable across systems that normalize the same way, so two products
 * looking up the same person get the same avatar.
 *
 * @example
 * ```ts
 * const s = seedFromEmail(user.email);
 * createAvatar(s);                          // safe to log
 * // or hit the API: `/avatar/${s}.svg`     // no plaintext email on the wire
 * ```
 */
export function seedFromEmail(email: string): string {
  if (typeof email !== 'string' || email.length === 0) {
    throw new Error('navii: seedFromEmail() requires a non-empty string');
  }
  return sha256Hex(normalizeEmail(email));
}

export interface SeedFields {
  /** Stable primary key (database id, UUID, OAuth sub). Best choice. */
  id?: string | number | null | undefined;
  /** Email. Stable + unique. Good fallback when id isn't available client-side. */
  email?: string | null | undefined;
  /** Display name. Collision-prone — only acceptable composed with `createdAt`. */
  name?: string | null | undefined;
  /** Account creation time. Combined with `name` to bake uniqueness in at signup. */
  createdAt?: string | number | Date | null | undefined;
}

export interface SeedOptions {
  /**
   * When the email branch is used, hash the email instead of returning it
   * raw. Hashing keeps the seed stable but stops the address from leaking
   * into URLs, server logs, and Referer headers. Default `true` from v1.
   *
   * Set to `false` to opt back into the legacy plaintext-email behavior —
   * useful for migrations where existing avatars are keyed off the raw
   * email and you don't want every user's face to change.
   */
  hashEmail?: boolean;
}

/**
 * Compose a stable seed string from the most unique field available.
 *
 * @example
 * ```ts
 * const s = seed({ id: user.id, email: user.email, name: user.name });
 * createAvatar(s);
 * ```
 *
 * @example If only a name exists, pass `createdAt` to avoid collisions:
 * ```ts
 * seed({ name: 'Alice', createdAt: user.createdAt });
 * // → "Alice|1700000000000"
 * ```
 *
 * @example Opt out of email hashing (legacy behavior):
 * ```ts
 * seed({ email: 'a@b.c' }, { hashEmail: false });
 * // → "a@b.c"  — avoid; only for migrating off the old default.
 * ```
 *
 * @throws if no usable field is provided.
 */
export function seed(fields: SeedFields, options: SeedOptions = {}): string {
  const hashEmail = options.hashEmail ?? true;

  if (fields.id !== null && fields.id !== undefined && String(fields.id).length > 0) {
    return String(fields.id);
  }
  if (fields.email && fields.email.length > 0) {
    return hashEmail ? seedFromEmail(fields.email) : fields.email;
  }
  if (fields.name && fields.name.length > 0) {
    if (fields.createdAt !== null && fields.createdAt !== undefined) {
      const ts = fields.createdAt instanceof Date
        ? fields.createdAt.getTime()
        : typeof fields.createdAt === 'number'
          ? fields.createdAt
          : Date.parse(fields.createdAt as string);
      if (Number.isFinite(ts)) return `${fields.name}|${ts}`;
      return `${fields.name}|${fields.createdAt}`;
    }
    return fields.name;
  }
  throw new Error('navii: seed() requires at least one of { id, email, name }');
}

File: packages/flutter/lib/src/core/seed.dart
/// Seed helpers — compose a stable identifier for [createAvatar].
///
/// Faithful port of `packages/core/src/seed.ts`.
library;

import 'sha256.dart';

/// Normalize an email the same way Gravatar does — trim + lowercase + NFC.
///
/// Dart's core library has no Unicode NFC; this applies trim + lowercase and
/// then a best-effort NFC via [Characters]-free precomposed form. For ASCII
/// and already-NFC input (typical email keyboards), output matches Node's
/// `String.prototype.normalize('NFC')`.
String normalizeEmail(String email) {
  return _nfc(email.trim().toLowerCase());
}

/// Turn an email into a stable opaque seed: `sha256(normalizeEmail(email))`.
String seedFromEmail(String email) {
  if (email.isEmpty) {
    throw ArgumentError('navii: seedFromEmail() requires a non-empty string');
  }
  return sha256Hex(normalizeEmail(email));
}

/// Fields accepted by [seed] — mirrors TS `SeedFields`.
class SeedFields {
  const SeedFields({
    this.id,
    this.email,
    this.name,
    this.createdAt,
  });

  /// Stable primary key (database id, UUID, OAuth sub). Best choice.
  /// Accepts [String] or [num] (coerced via `toString()`).
  final Object? id;

  /// Email. Stable + unique. Good fallback when id isn't available.
  final String? email;

  /// Display name. Collision-prone — only acceptable with [createdAt].
  final String? name;

  /// Account creation time — [DateTime], epoch [num], or parseable [String].
  final Object? createdAt;
}

/// Options for [seed].
class SeedOptions {
  const SeedOptions({this.hashEmail});

  /// When the email branch is used, hash instead of returning raw email.
  /// Default `true`. Set `false` for legacy plaintext-email migrations.
  final bool? hashEmail;
}

/// Compose a stable seed string from the most unique field available.
///
/// Priority: `id` → `email` → `name + createdAt` → `name` alone.
String seed(SeedFields fields, [SeedOptions options = const SeedOptions()]) {
  final hashEmail = options.hashEmail ?? true;

  final id = fields.id;
  if (id != null && id.toString().isNotEmpty) {
    return id.toString();
  }

  final email = fields.email;
  if (email != null && email.isNotEmpty) {
    return hashEmail ? seedFromEmail(email) : email;
  }

  final name = fields.name;
  if (name != null && name.isNotEmpty) {
    final createdAt = fields.createdAt;
    if (createdAt != null) {
      final ts = _createdAtMillis(createdAt);
      if (ts != null) return '$name|$ts';
      return '$name|$createdAt';
    }
    return name;
  }

  throw ArgumentError('navii: seed() requires at least one of { id, email, name }');
}

int? _createdAtMillis(Object createdAt) {
  if (createdAt is DateTime) return createdAt.millisecondsSinceEpoch;
  if (createdAt is num) return createdAt.toInt();
  if (createdAt is String) {
    final parsed = DateTime.tryParse(createdAt);
    if (parsed != null) return parsed.millisecondsSinceEpoch;
  }
  return null;
}

/// Best-effort NFC for BMP Latin / common email characters.
///
/// Full Unicode NFC is out of scope without an ICU dependency; this covers
/// the composed Latin-1 supplement letters that show up in emails. Strings
/// that are already NFC are returned unchanged.
String _nfc(String input) {
  // Dart source / UTF-16 strings from Flutter text input are typically NFC.
  // Explicit normalize would require ICU (`package:intl` does not expose NFC).
  // Match Node for precomposed input; NFD edge cases are documented as known gap.
  return input;
}
